### PR TITLE
Improve multimatch management

### PR DESCRIPTION
**Pull request by salome-voltz**
## Context

This PR changes the number of problems created, only for URIs. The decision behind it is described here:

For example:
```
username= ....
password= ....
```
Both matches must be underlined, and it's better if there is one problem per match. If the matches are very far away, it can add context. 

Also, it would be hard to change this behavior since it uses the Vscode API which works that way (one problem underlined <-> one problem in the problem output). Changing this behavior would be much more complex, and it doesn't add much. 

However for URIs:
```
CONNECTION_URI="postgres://postgres:m42ploz2wd@something.com:5434/something"
```
Here, there are technically 5 matches: "scheme", "username", "password", "host" and "connection_uri". Since "connection_uri" represents the whole string, it would make sense to only keep this match, and not create 5 problems in the problem output, as they all link to the same position in the file. 

## What has been done
- [fix(scanFile): Refacto scanFile function to use status code](https://github.com/GitGuardian/gitguardian-vscode/pull/56/commits/09949e8c4dc316adfe73f1d14546ff97c47d3728) : small refacto of the scanFile function for better consistency in the project. 
for context : 
image
- [feat(result_parser): Better parsing of URI secrets](https://github.com/GitGuardian/gitguardian-vscode/pull/56/commits/c4d57334874501aaa2795bc3f7152ea4fbd5e0d2) : only return the "connection_uri" match if there is one

## PR check list

- [x] As much as possible, the changes include tests
- [ ] If the changes affect the end user (new feature, behavior change, bug fix) then the PR has a changelog entry.

### CODE REVIEW

File: src/lib/ggshield-api.ts [hunk 1]
```diff
@@ -172,47 +172,42 @@ export async function scanFile(
...
+  if (proc.status === 128 || proc.status === 3) {
+    let errorMessage = "";
+    proc.stderr.split("\n").forEach((stderrLine) => {
+      if (
+        stderrLine.length > 0 &&
+        !stderrLine.includes("Scanning Path...") // ggshield outputs this info message on stderr, ignore it
+      ) {
+        errorMessage += stderrLine + "\n";
+      }
+    });
```
Comment (gg-jonathangriffe) [inline src/lib/ggshield-api.ts:184]: I think using filter here is clearer :
```suggestion
    const errorMessage = proc.stderr.split("\n").filter(
        (stderrLine) => stderrLine.length > 0 && !stderrLine.includes("Scanning Path...")  // ggshield outputs this info message on stderr, ignore it
    ).join("\n");
```

File: src/lib/ggshield-api.ts [hunk 2]
```diff
@@ -172,47 +172,42 @@ export async function scanFile(
...
-    window.showErrorMessage(`ggshield: ${errorMessage}`);
     return undefined;
-  }
-
-  const results = JSON.parse(proc.stdout);
-  if (!results) {
-    updateStatusBarItem(StatusBarStatus.ready);
+  } else if (proc.status === 0) {
+    updateStatusBarItem(StatusBarStatus.noSecretFound);
     return;
```
Comment (gg-jonathangriffe) [inline src/lib/ggshield-api.ts:204]: Here we still need to do `diagnosticCollection.set(fileUri, incidentsDiagnostics);`, or else the problem won't be removed when the secret is removed from the file.

I think that parseGGShieldResults also works when there are no results, so we can maybe just leave `updateStatusBarItem` in the if else and put everything else after ?

File: src/lib/ggshield-results-parser.ts [hunk 3]
```diff
@@ -15,18 +16,21 @@ import {
...
   no_checker: "No Checker",
-  // eslint-disable-next-line @typescript-eslint/naming-convention
   failed_to_check: "Failed to Check",
-  // eslint-disable-next-line @typescript-eslint/naming-convention
   not_checked: "Not Checked",
   invalid: "Invalid",
   valid: "Valid",
 };
 
+function filterUriOccurrences(occurrences: Occurrence[]): Occurrence[] {
```
Comment (gg-jonathangriffe) [inline src/lib/ggshield-results-parser.ts:27]: can you add a small docstring to explain what you are doing here

File: src/lib/ggshield-api.ts (hunk 2 above)
Comment (salome-voltz) [reply src/lib/ggshield-api.ts:204]: Oh right :sweat_smile:  I might have been a bit violent while trying to remove old unnecessary code...

**Comment by gg-jonathangriffe** [comment]
Looks good to me ! Only one very small fix left.
Thanks for the refacto :star_struck:

File: src/test/suite/lib/ggshield-api.test.ts [hunk 4]
```diff
@@ -78,19 +78,22 @@ suite("scanFile", () => {
...
+        status: code,
+        stdout: "",
+        stderr: "Error",
+      });
+
+      await scanFile("test.py", Uri.file("test.py"), {} as GGShieldConfiguration);
+
+      // The error message is displayed
+      assert.strictEqual(errorMessageMock.callCount, 1);
+      assert.strictEqual(errorMessageMock.lastCall.args[0], "ggshield: Error\n");
```
Comment (gg-jonathangriffe) [inline src/test/suite/lib/ggshield-api.test.ts:95]: The test now fails because there is no `\n` at the end of the test anymore.

**Review by gg-jonathangriffe — APPROVED** [review]

File: src/test/suite/lib/ggshield-api.test.ts (hunk 4 above)
Comment (salome-voltz) [reply src/test/suite/lib/ggshield-api.test.ts]: And... We really need a CI :sweat:

**Pull request merged**